src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php

```diff
@@ -72,23 +72,49 @@
  */
 abstract class AnnotationClassLoader implements LoaderInterface
 {
+    /**
+     * @var Reader|null
+     * @deprecated in Symfony 6.3, this property will be removed in Symfony 7.
+     */
     protected $reader;
+
+    /**
+     * @var string|null
+     * @internal since Symfony 6.3, this property will be private in Symfony 7.
+     */
     protected $env;
 
     /**
      * @var string
+     * @internal since Symfony 6.3, this property will be private in Symfony 7.
      */
     protected $routeAnnotationClass = RouteAnnotation::class;
 
     /**
      * @var int
+     * @internal since Symfony 6.3, this property will be private in Symfony 7.
      */
     protected $defaultRouteIndex = 0;
 
-    public function __construct(Reader $reader = null, string $env = null)
+    private bool $hasDeprecatedAnnotations = false;
+
+    /**
+     * @param string|null $env
+     */
+    public function __construct($env = null)
     {
-        $this->reader = $reader;
-        $this->env = $env;
+        if ($env instanceof Reader || func_num_args() > 1 && null !== func_get_arg(1)) {
+            trigger_deprecation( 'symfony/routing', '6.3', 'Passing an instance of "%s" as first and the environment as second argument to "%s" is deprecated. Pass the environment as first argument instead.', Reader::class, __METHOD__);
```

Comment:
Given that FrameworkBundle cannot stop passing a Reader to that loader until 7.0 (otherwise annotations would stop working immediately), wouldn't we end up triggering this deprecation all the time ?

Reply:
We should not trigger a deprecation here when annotations are enabled in FrameworkBundle (which is not even a deprecated setting yet). We should have deprecations only when you actually have routing annotations being used, at least in the case of the loader configured by FrameworkBundle. Otherwise, projects will not be able to resolve their deprecations (and they will report that as a bug in Symfony, and happened in the past for similar cases)

Reply:
I think it will become a deprecated setting eventually, but as long as it's not, @stof is right. The app should be allowed to enable annotations in the framework and as long as routing is not done via annotations, no deprecation should be triggered.

Maybe we could add a switch to disable annotations for routing only and remove it again should we manage to deprecate annotations support everywhere before 6.3.